File: src/programmers/json/JsonValidateParseProgrammer.ts
import ts from "typescript";

import { IdentifierFactory } from "../../factories/IdentifierFactory";
import { StatementFactory } from "../../factories/StatementFactory";
import { TypeFactory } from "../../factories/TypeFactory";

import { IProgrammerProps } from "../../transformers/IProgrammerProps";
import { ITypiaContext } from "../../transformers/ITypiaContext";

import { FeatureProgrammer } from "../FeatureProgrammer";
import { ValidateProgrammer } from "../ValidateProgrammer";
import { FunctionProgrammer } from "../helpers/FunctionProgrammer";

export namespace JsonValidateParseProgrammer {
  export const decompose = (props: {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    functor: FunctionProgrammer;
    type: ts.Type;
    name: string | undefined;
  }): FeatureProgrammer.IDecomposed => {
    const validate: FeatureProgrammer.IDecomposed =
      ValidateProgrammer.decompose({
        ...props,
        context: {
          ...props.context,
          options: {
            ...props.context.options,
            functional: false,
            numeric: false,
          },
        },
        config: {
          equals: false,
        },
      });
    return {
      functions: validate.functions,
      statements: [
        ...validate.statements,
        StatementFactory.constant({
          name: "__validate",
          value: validate.arrow,
        }),
      ],
      arrow: ts.factory.createArrowFunction(
        undefined,
        undefined,
        [IdentifierFactory.parameter("input", TypeFactory.keyword("string"))],
        props.context.importer.type({
          file: "typia",
          name: "IValidation",
          arguments: [
            props.context.importer.type({
              file: "typia",
              name: "Primitive",
              arguments: [
                ts.factory.createTypeReferenceNode(
                  props.name ??
                    TypeFactory.getFullName({
                      checker: props.context.checker,
                      type: props.type,
                    }),
                ),
              ],
            }),
          ],
        }),
        undefined,
        ts.factory.createAsExpression(
          ts.factory.createCallExpression(
            ts.factory.createIdentifier("__validate"),
            undefined,
            [
              ts.factory.createCallExpression(
                ts.factory.createIdentifier("JSON.parse"),
                undefined,
                [ts.factory.createIdentifier("input")],
              ),
            ],
          ),
          ts.factory.createTypeReferenceNode("any"),
        ),
      ),
    };
  };

  export const write = (props: IProgrammerProps): ts.CallExpression => {
    const functor: FunctionProgrammer = new FunctionProgrammer(
      props.modulo.getText(),
    );
    const result: FeatureProgrammer.IDecomposed = decompose({
      context: props.context,
      modulo: props.modulo,
      functor,
      type: props.type,
      name: props.name,
    });
    return FeatureProgrammer.writeDecomposed({
      modulo: props.modulo,
      functor,
      result,
    });
  };
}

File: packages/typia/native/core/programmers/json/JsonValidateParseProgrammer.go
package json

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  shimchecker "github.com/microsoft/typescript-go/shim/checker"
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers"
  nativehelpers "github.com/samchon/typia/packages/typia/native/core/programmers/helpers"
  nativeinternal "github.com/samchon/typia/packages/typia/native/core/programmers/internal"
)

type jsonValidateParseProgrammerNamespace struct{}

var JsonValidateParseProgrammer = jsonValidateParseProgrammerNamespace{}

type JsonValidateParseProgrammer_DecomposeProps struct {
  Context nativecontext.ITypiaContext
  Modulo  *shimast.Node
  Functor *nativehelpers.FunctionProgrammer
  Type    *shimchecker.Type
  Name    *string
}

func (jsonValidateParseProgrammerNamespace) Decompose(props JsonValidateParseProgrammer_DecomposeProps) nativeinternal.FeatureProgrammer_IDecomposed {
  validate := nativeprogrammers.ValidateProgrammer.Decompose(nativeprogrammers.ValidateProgrammer_DecomposeProps{
    Context: jsonParseProgrammer_context(props.Context, &jsonParseProgrammer_options{
      Functional: jsonParseProgrammer_bool(false),
      Numeric:    jsonParseProgrammer_bool(false),
    }),
    Modulo:  props.Modulo,
    Functor: props.Functor,
    Config:  nativeprogrammers.ValidateProgrammer_IConfig{Equals: false},
    Type:    props.Type,
    Name:    props.Name,
  })

  return nativeinternal.FeatureProgrammer_IDecomposed{
    Functions: validate.Functions,
    Statements: append(append([]*shimast.Node{}, validate.Statements...), nativefactories.StatementFactory.Constant(nativefactories.StatementFactory_ConstantProps{
      Name:  "__validate",
      Value: validate.Arrow,
    })),
    Arrow: jsonParseProgrammer_factory.NewArrowFunction(
      nil,
      nil,
      jsonParseProgrammer_factory.NewNodeList([]*shimast.Node{
        nativefactories.IdentifierFactory.Parameter("input", nativefactories.TypeFactory.Keyword("string"), nil),
      }),
      jsonProgrammer_import_type(props.Context, nativeprogrammers.ImportProgrammer_TypeProps{
        File: "typia",
        Name: "IValidation",
        Arguments: []*shimast.TypeNode{
          jsonProgrammer_import_type(props.Context, nativeprogrammers.ImportProgrammer_TypeProps{
            File: "typia",
            Name: "Primitive",
            Arguments: []*shimast.TypeNode{
              jsonProgrammer_typeReference(props.Context, props.Type, props.Name),
            },
          }),
        },
      }),
      nil,
      jsonParseProgrammer_factory.NewToken(shimast.KindEqualsGreaterThanToken),
      jsonParseProgrammer_factory.NewAsExpression(
        jsonParseProgrammer_factory.NewCallExpression(
          jsonParseProgrammer_factory.NewIdentifier("__validate"),
          nil,
          nil,
          jsonParseProgrammer_factory.NewNodeList([]*shimast.Node{
            jsonParseProgrammer_factory.NewCallExpression(
              jsonParseProgrammer_factory.NewIdentifier("JSON.parse"),
              nil,
              nil,
              jsonParseProgrammer_factory.NewNodeList([]*shimast.Node{
                jsonParseProgrammer_factory.NewIdentifier("input"),
              }),
              shimast.NodeFlagsNone,
            ),
          }),
          shimast.NodeFlagsNone,
        ),
        jsonParseProgrammer_factory.NewTypeReferenceNode(jsonParseProgrammer_factory.NewIdentifier("any"), nil),
      ),
    ),
  }
}

func (jsonValidateParseProgrammerNamespace) Write(props nativecontext.IProgrammerProps) *shimast.Node {
  method := ""
  if props.Modulo != nil {
    method = props.Modulo.Text()
  }
  functor := nativehelpers.NewFunctionProgrammer(method)
  result := JsonValidateParseProgrammer.Decompose(JsonValidateParseProgrammer_DecomposeProps{
    Context: props.Context,
    Modulo:  props.Modulo,
    Functor: functor,
    Type:    props.Type,
    Name:    props.Name,
  })
  return nativeinternal.FeatureProgrammer.WriteDecomposed(nativeinternal.FeatureProgrammer_WriteDecomposedProps{
    Modulo:  props.Modulo,
    Functor: functor,
    Result:  result,
  })
}
